File: services/configd/tests/support/projections/derive-projection-id.ts
import {
  createHash
} from "node:crypto";
import type {
  ConsumerBinding
} from "../../generated/v1/configd.js";

const alphabet = "abcdefghijklmnopqrstuvwxyz234567";

export function deriveProjectionId(
  kind: "configuration" | "secret",
  binding: ConsumerBinding
): string {
  const chunks = [
    Buffer.from("ctlflow.configuration.v1.Projection", "ascii"),
    Buffer.of(0),
    Buffer.of(kind === "configuration" ? 1 : 2),
    field(binding.placement?.placementId ?? "")
  ];
  const placement = binding.placement;
  if (placement?.global !== undefined) {
    chunks.push(Buffer.of(1));
  } else if (placement?.tenant !== undefined) {
    chunks.push(
      Buffer.of(2),
      field(placement.tenant.tenantId));
  } else if (placement?.workspace !== undefined) {
    chunks.push(
      Buffer.of(3),
      field(placement.workspace.tenantId),
      field(placement.workspace.workspaceId));
  } else if (placement?.user !== undefined) {
    chunks.push(
      Buffer.of(4),
      field(placement.user.tenantId),
      field(placement.user.accountPrincipalId));
  } else {
    throw new Error("Projection binding has no scope");
  }
  chunks.push(
    field(binding.consumerId),
    field(binding.purpose));
  return `prj_${base32(createHash("sha256")
    .update(Buffer.concat(chunks))
    .digest())}`;
}

function field(value: string): Buffer {
  const bytes = Buffer.from(value, "utf8");
  const length = Buffer.alloc(4);
  length.writeUInt32BE(bytes.length);
  return Buffer.concat([length, bytes]);
}

function base32(value: Buffer): string {
  let buffer = 0;
  let bits = 0;
  let result = "";
  for (const byte of value) {
    buffer = (buffer << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      result += alphabet[(buffer >> bits) & 31];
    }
  }
  if (bits > 0) {
    result += alphabet[(buffer << (5 - bits)) & 31];
  }
  return result;
}

File: services/configd/csharp/src/CtlFlow.Configuration.Configd.Domain/Projections/DeriveProjectionId.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;
using CtlFlow.Configuration.Configd.Domain.Bindings;
using CtlFlow.Configuration.Configd.Domain.Identifiers;

namespace CtlFlow.Configuration.Configd.Domain.Projections;

public static partial class ProjectionIdentities
{
    private const string Domain = "ctlflow.configuration.v1.Projection";
    private const string Base32Alphabet = "abcdefghijklmnopqrstuvwxyz234567";

    public static ValueTask<ProjectionId> DeriveProjectionId(
        ProjectionDataKind kind,
        ConsumerBinding binding,
        CancellationToken cancellation)
    {
        cancellation.ThrowIfCancellationRequested();
        using var stream = new MemoryStream(512);
        stream.Write(Encoding.ASCII.GetBytes(Domain));
        stream.WriteByte(0);
        stream.WriteByte(kind switch
        {
            ProjectionDataKind.Configuration => 1,
            ProjectionDataKind.Secret => 2,
            _ => throw new InvalidOperationException(
                "Projection data kind is invalid")
        });
        WriteField(stream, binding.Placement.PlacementId.Value);
        switch (binding.Placement.Scope)
        {
            case PlacementScope.Global:
                stream.WriteByte(1);
                break;
            case PlacementScope.Tenant tenant:
                stream.WriteByte(2);
                WriteField(stream, tenant.TenantId.Value);
                break;
            case PlacementScope.Workspace workspace:
                stream.WriteByte(3);
                WriteField(stream, workspace.TenantId.Value);
                WriteField(stream, workspace.WorkspaceId.Value);
                break;
            case PlacementScope.User user:
                stream.WriteByte(4);
                WriteField(stream, user.TenantId.Value);
                WriteField(stream, user.AccountPrincipalId.Value);
                break;
            default:
                throw new InvalidOperationException(
                    "Placement scope is invalid");
        }

        WriteField(stream, binding.ConsumerId.Value);
        WriteField(stream, binding.Purpose.Value);
        var digest = SHA256.HashData(stream.GetBuffer().AsSpan(
            0,
            checked((int)stream.Length)));
        return ValueTask.FromResult(
            ProjectionId.FromDigest(EncodeBase32(digest)));
    }

    private static void WriteField(Stream stream, string value)
    {
        var bytes = Encoding.UTF8.GetBytes(value);
        Span<byte> length = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(
            length,
            checked((uint)bytes.Length));
        stream.Write(length);
        stream.Write(bytes);
    }

    private static string EncodeBase32(ReadOnlySpan<byte> value)
    {
        Span<char> encoded = stackalloc char[52];
        var buffer = 0;
        var bits = 0;
        var output = 0;
        foreach (var item in value)
        {
            buffer = (buffer << 8) | item;
            bits += 8;
            while (bits >= 5)
            {
                bits -= 5;
                encoded[output++] =
                    Base32Alphabet[(buffer >> bits) & 31];
            }
        }

        if (bits > 0)
        {
            encoded[output++] =
                Base32Alphabet[(buffer << (5 - bits)) & 31];
        }

        if (output != encoded.Length)
        {
            throw new InvalidOperationException(
                "Projection digest encoding failed");
        }

        return new string(encoded);
    }
}
